### ISSUE TITLE
Is there MC sampling in the training phase?

### ISSUE THREAD

**Issue by huiminzeng** [y0ast/DUE#4]
Dear authors,

I really like the idea of this paper. Thank you very much! However, while reading the code, I am a little confused by the implementation of DUE.  

In the paper (Appendix A.2, right after Equation 9), it is mentioned that MC sampling is used for training the classification model. But according to the implementation, I couldn't find the location where MC sampling is implemented. Based on my observation of the code, the output of the model is a Gaussian distribution. Then this distribution is fed into the VariationalElbo directly to compute the loss. Therefore, could you please link me to the place where MC sampling is implemented?

Another question about MC sampling during training is more fundamental. I think MC sampling is non-differentiable. If MC sampling exists during training, how the gradient could be propagated to the feature extractor part?

Please let me apologize if I am asking wrong questions. (I am new to Gaussian process.) 

Thank you very much

**Comment by y0ast** [comment]
Hello! Thank you for your interested in the paper.

MC sampling happens inside the likelihood object: https://github.com/y0ast/DUE/blob/main/train_due.py#L89 (which is passed to the ELBO, and called internally during training).

You can change the number of samples using this feature:
https://docs.gpytorch.ai/en/stable/settings.html#gpytorch.settings.num_likelihood_samples

Like we do here in the evaluation: https://github.com/y0ast/DUE/blob/main/lib/evaluate_ood.py#L45-L47

> MC sampling non differentiable

Yes this is a bit tricky. We actually sample from normal distributions and push through the softmax. This is differentiable (through differentiating the softmax and the reparametrization trick). We don't sample from a categorical distribution, which would indeed be hard to differentiate through.
